File: spotlight-sarajevo-be/src/main/java/com/spotlightsarajevo/modules/guide/service/TouristGuideService.java
package com.spotlightsarajevo.modules.guide.service;

import com.spotlightsarajevo.modules.guide.api.dto.*;
import org.apache.catalina.connector.Response;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.ResponseEntity;

import java.security.Principal;
import java.util.List;

/**
 * TouristGuideService interface defines methods for managing tourist guides.
 * It includes methods to find all guides, find guides by category,
 */
public interface TouristGuideService {
    /**
     * Retrieves a list of all tourist guides in shorthand format.
     *
     * @return ResponseEntity containing a list of TouristGuideShorthandModel
     */
    ResponseEntity<List<TouristGuideShorthandModel>> findAllGuides();

    /**
     * Retrieves a list of paginated tourist guides in shorthand format
     *
     * @return A list of paginated response, sorted and based on the search term
     */
    ResponseEntity<Page<TouristGuideShorthandModel>> findGuidesPaginated(PageRequest request, String searchTerm, String sortOption);

    /**
     * Retrieves a list of tourist guides by category ID in shorthand format.
     *
     * @param categoryId the unique identifier of the category
     * @return ResponseEntity containing a list of TouristGuideShorthandModel
     */
    ResponseEntity<List<TouristGuideShorthandModel>> findByCategory(Integer categoryId);

    /**
     * Retrieves an overview of a specific tourist guide by its slug.
     *
     * @param slug the unique slug identifier of the tourist guide
     * @return ResponseEntity containing the TouristGuideOverviewModel
     */
    ResponseEntity<TouristGuideOverviewModel> findGuideOverview(String slug);

    /**
     * Creates a new tourist guide based on the provided TouristGuideCreateModel.
     *
     * @param request the model containing information to create a new tourist guide
     * @param principal the security principal representing the current user
     * @return ResponseEntity containing the created TouristGuideOverviewModel
     */
    ResponseEntity<TouristGuideOverviewModel> create(TouristGuideCreateModel request, Principal principal);

    /**
     * Updates an existing tourist guide, alongside its individual sections.
     *
     * @param request The TouristGuideUpdateModel sent from the FE
     * @param principal The principal object that is sent alongside the authenticated user data
     * @return ResponseEntity containing the newly updated TouristGuideModel (The simple one with the updated status)
     */
    ResponseEntity<TouristGuideModel> update(TouristGuideUpdateModel request, Principal principal);

    /**
     * Deletes a tourist guide by its ID.
     *
     * @param id the unique identifier of the tourist guide to be deleted
     * @return ResponseEntity containing the deleted TouristGuideModel
     */
    ResponseEntity<TouristGuideModel> delete(Integer id);
}

File: spotlight-sarajevo-fe/src/app/services/tourist.guide.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { TouristGuideCreateModel, TouristGuideModel, TouristGuideOverviewModel, TouristGuideShorthandModel, TouristGuideUpdateModel } from '../shared/models/tourist.guide.model';
import { environment } from '../../environments/environment';
import { PageResponseModel } from '../shared/models/shared.model';

/**
 * Service for interacting with the Tourist Guide API.
 * Provides methods to fetch tourist guides and create new ones.
 * All requests include credentials for authentication.
 * 
 * Models and entities incorporated in the methods: TouristGuideShorthandModel, TouristGuideOverviewModel, TouristGuideCreateModel
 * 
 * @version 1.0.0
 * @author hajrudin.imamovic
 */
@Injectable({
  providedIn: 'root'
})
export class TouristGuideService {
  private API_URL = environment.API_URL + '/guide';

  constructor(private http: HttpClient) {}

  /**
   * Fetches all tourist guides in shorthand format
   */
  findAllGuides(): Observable<TouristGuideShorthandModel[]> {
    return this.http.get<TouristGuideShorthandModel[]>(`${this.API_URL}/all`, {
      withCredentials: true
    });
  }

  findGuidesPaginated(
    pageNumber: number, pageSize: number, searchTerm: string, sortOption: string
  ) {
    return this.http.get<PageResponseModel<TouristGuideShorthandModel>>(`${this.API_URL}?pageNumber=${pageNumber}&pageSize=${pageSize}&searchTerm=${searchTerm}&sortOption=${sortOption}`, {
      withCredentials: true
    })
  }

  /**
   * Fetches guides based on a specific category ID
   * @param categoryId The ID of the category to filter guides by
   */
  findGuidesByCategory(categoryId: number): Observable<TouristGuideShorthandModel[]> {
    return this.http.get<TouristGuideShorthandModel[]>(`${this.API_URL}/category/${categoryId}`, {
      withCredentials: true
    });
  }

  /**
   * Fetches a specific guide overview by its slug
   * @param slug The unique string identifier for the guide
   */
  findGuideOverview(slug: string): Observable<TouristGuideOverviewModel> {
    return this.http.get<TouristGuideOverviewModel>(`${this.API_URL}/${slug}`, {
      withCredentials: true
    });
  }

  /**
   * Creates a new tourist guide
   * @param request The guide creation data
   */
  createGuide(request: TouristGuideCreateModel): Observable<TouristGuideOverviewModel> {
    return this.http.post<TouristGuideOverviewModel>(`${this.API_URL}/create`, request, {
      withCredentials: true
    });
  }

  /**
   * Updates an existing tourist guide
   * @param guideId The ID of the guide to update
   * @param request The updated guide data
   */
  updateGuide(request: TouristGuideUpdateModel) {
    return this.http.put<TouristGuideModel>(`${this.API_URL}`, request, {
      withCredentials: true
    });
   }

   /**
    * Deletes a tourist guide by its ID
    * @param guideId The ID of the guide to delete
    */
   deleteGuide(guideId: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${guideId}`, {
      withCredentials: true
    });
   }
}